File: projects/composablekernel/dispatcher/examples/grouped_conv/cpp/04_registry_json.cpp
// Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
// SPDX-License-Identifier: MIT

/**
 * Example 04: Multi-Registry and JSON Export
 *
 * Demonstrates:
 * - Multiple registries for different workloads (throughput vs latency)
 * - GroupedConvDispatcher for kernel selection
 * - JSON export with statistics
 * - filter_by_arch for architecture-specific deployment
 *
 * Build: cd dispatcher/build && cmake .. && make grouped_conv_04_registry_json
 */

#include <iostream>
#include <iomanip>
#include <vector>

#include "ck_tile/dispatcher/grouped_conv_utils.hpp"
#include "ck_tile/dispatcher/example_args.hpp"

using namespace ck_tile::dispatcher;
using namespace ck_tile::dispatcher::grouped_conv_utils;
using GroupedConvSig  = grouped_conv_decl::GroupedConvSignature;
using GroupedConvAlgo = grouped_conv_decl::GroupedConvAlgorithm;

// Throughput-optimized kernels (large tiles)
DECL_GROUPED_CONV_KERNEL_SET(
    throughput_kernels,
    .add(GroupedConvSig().dtype("fp16").layout("nhwc").conv_type("forward").dims(2),
         GroupedConvAlgo().tile(1, 256, 256).pipeline("compv4").vector_sizes(4, 8, 8),
         "gfx950")
    .add(GroupedConvSig().dtype("fp16").layout("nhwc").conv_type("forward").dims(2),
         GroupedConvAlgo().tile(1, 128, 256).pipeline("compv4").vector_sizes(4, 8, 8),
         "gfx950"));

// Latency-optimized kernels (small tiles)
DECL_GROUPED_CONV_KERNEL_SET(
    latency_kernels,
    .add(GroupedConvSig().dtype("fp16").layout("nhwc").conv_type("forward").dims(2),
         GroupedConvAlgo().tile(1, 64, 64).pipeline("compv3").vector_sizes(4, 8, 8),
         "gfx950")
    .add(GroupedConvSig().dtype("fp16").layout("nhwc").conv_type("forward").dims(2),
         GroupedConvAlgo().tile(1, 32, 32).pipeline("compv3").vector_sizes(4, 4, 4),
         "gfx950"));

// Multi-arch kernels
DECL_GROUPED_CONV_KERNEL_SET(
    multi_arch_kernels,
    .add(GroupedConvSig().dtype("fp16").layout("nhwc").conv_type("forward").dims(2),
         GroupedConvAlgo().tile(1, 128, 128).pipeline("compv4"),
         "gfx950")
    .add(GroupedConvSig().dtype("fp16").layout("nhwc").conv_type("forward").dims(2),
         GroupedConvAlgo().tile(1, 128, 128).pipeline("compv3"),
         "gfx942")
    .add(GroupedConvSig().dtype("bf16").layout("nhwc").conv_type("forward").dims(2),
         GroupedConvAlgo().tile(1, 128, 128).pipeline("compv4"),
         "gfx950"));

int main(int argc, char* argv[])
{
    utils::ExampleArgs args("Example 04: Multi-Registry & JSON Export",
                            "Separate registries and JSON export with statistics");
    args.add_option("--output", "", "JSON output file (optional)");

    if(!args.parse(argc, argv))
        return 0;

    utils::print_header("Example 04: Multi-Registry & JSON Export");

    auto& kset_reg = GroupedConvKernelSetRegistry::instance();

    // =========================================================================
    // Throughput registry
    // =========================================================================
    std::cout << "\n--- Throughput Registry ---\n";
    GroupedConvRegistry throughput_reg;
    throughput_reg.set_name("throughput");
    throughput_reg.register_set(kset_reg.get("throughput_kernels"), GroupedConvRegistry::Priority::High);
    std::cout << "  Kernels: " << throughput_reg.size() << "\n";

    // =========================================================================
    // Latency registry
    // =========================================================================
    std::cout << "\n--- Latency Registry ---\n";
    GroupedConvRegistry latency_reg;
    latency_reg.set_name("latency");
    latency_reg.register_set(kset_reg.get("latency_kernels"), GroupedConvRegistry::Priority::High);
    std::cout << "  Kernels: " << latency_reg.size() << "\n";

    // =========================================================================
    // Dispatcher selection
    // =========================================================================
    std::cout << "\n--- Dispatcher Selection ---\n";

    auto large_problem = create_grouped_conv2d_problem(8, 128, 256, 56, 56, 3, 3, 1, 1);
    auto small_problem = create_grouped_conv2d_problem(1, 32, 64, 14, 14, 1, 1, 1, 0);

    GroupedConvDispatcher throughput_disp(&throughput_reg);
    GroupedConvDispatcher latency_disp(&latency_reg);

    auto* tp_sel = throughput_disp.select(large_problem);
    auto* lt_sel = latency_disp.select(small_problem);

    std::cout << "  Large problem -> throughput: " << (tp_sel ? tp_sel->name() : "none") << "\n";
    std::cout << "  Small problem -> latency:    " << (lt_sel ? lt_sel->name() : "none") << "\n";

    // =========================================================================
    // Multi-arch with filter_by_arch
    // =========================================================================
    std::cout << "\n--- Multi-Arch Filter ---\n";
    GroupedConvRegistry multi_arch_reg;
    multi_arch_reg.set_name("multi_arch");
    multi_arch_reg.register_set(kset_reg.get("multi_arch_kernels"));
    std::cout << "  Before filter: " << multi_arch_reg.size() << " kernels\n";

    auto removed = multi_arch_reg.filter_by_arch("gfx950");
    std::cout << "  Removed " << removed << " non-gfx950 kernels\n";
    std::cout << "  After filter:  " << multi_arch_reg.size() << " kernels\n";

    // =========================================================================
    // JSON export with statistics
    // =========================================================================
    std::cout << "\n--- JSON Export ---\n";

    // Merge all into one registry for comprehensive export
    GroupedConvRegistry combined;
    combined.set_name("all_conv_kernels");
    combined.register_set(kset_reg.get("throughput_kernels"));
    combined.register_set(kset_reg.get("latency_kernels"));
    combined.register_set(kset_reg.get("multi_arch_kernels"));

    std::string json = combined.export_json(true);
    std::cout << "  Total kernels in combined registry: " << combined.size() << "\n";
    std::cout << "  JSON size: " << json.size() << " bytes\n";

    // Print first portion
    std::cout << "\n  Preview:\n";
    auto preview = json.substr(0, std::min(json.size(), size_t(500)));
    std::cout << preview << "\n  ...\n";

    // Optionally write to file
    std::string output_file = args.get("--output", "");
    if(!output_file.empty())
    {
        combined.export_json_to_file(output_file, true);
        std::cout << "\n  Written to: " << output_file << "\n";
    }

    // =========================================================================
    // Summary
    // =========================================================================
    utils::print_separator();
    std::cout << "MULTI-REGISTRY & JSON FEATURES:\n";
    std::cout << "  - Separate registries: throughput vs latency\n";
    std::cout << "  - Priority-based kernel registration\n";
    std::cout << "  - GroupedConvDispatcher selects best kernel per problem\n";
    std::cout << "  - filter_by_arch() for deployment-time arch filtering\n";
    std::cout << "  - export_json(include_statistics=true) for analysis\n";
    utils::print_separator();

    std::cout << "\n  Status: PASS\n";
    return 0;
}

File: projects/composablekernel/dispatcher/examples/grouped_conv/python/04_registry_json.py
#!/usr/bin/env python3

# Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
# SPDX-License-Identifier: MIT

"""
Example 04: Registry & JSON Export/Import with GPU Execution

Demonstrates kernel registry management, JSON serialization, and GPU dispatch.

Usage:
    python3 04_registry_json.py
"""

import sys
import json
import numpy as np
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent.parent.parent / "python"))

from grouped_conv_utils import (
    GroupedConvKernelConfig,
    GroupedConvProblem,
    GroupedConvRegistry,
    GpuGroupedConvRunner,
    validate_grouped_conv_config,
    detect_gpu_arch,
)


def main():
    arch = detect_gpu_arch()
    print("=" * 70)
    print("Example 04: Registry & JSON Export/Import")
    print("=" * 70)
    print(f"\n  Arch: {arch}")

    # Step 1: Build throughput registry (large tiles)
    print("\n--- Step 1: Throughput Registry (large tiles) ---")
    tp_reg = GroupedConvRegistry("throughput")
    for variant in ["forward", "bwd_data", "bwd_weight"]:
        tp_reg.add(GroupedConvKernelConfig(
            variant=variant, ndim_spatial=2, arch=arch,
            tile_n=256, tile_k=256, pipeline="compv3",
        ))
    tp_reg.print_registry()

    # Step 2: Build latency registry (small tiles)
    print("\n--- Step 2: Latency Registry (small tiles) ---")
    lat_reg = GroupedConvRegistry("latency")
    for variant in ["forward", "bwd_data", "bwd_weight"]:
        lat_reg.add(GroupedConvKernelConfig(
            variant=variant, ndim_spatial=2, arch=arch,
            tile_n=64, tile_k=64, pipeline="compv3",
        ))
    lat_reg.print_registry()

    # Step 3: JSON export
    print("\n--- Step 3: JSON Export ---")
    combined = GroupedConvRegistry("all_conv_kernels")
    for k in tp_reg.kernels:
        combined.add(k)
    for k in lat_reg.kernels:
        combined.add(k)

    json_str = combined.to_json()
    print(f"  Combined: {len(combined)} kernels")
    print(f"  JSON size: {len(json_str)} bytes")
    print(f"  Preview:\n{json_str[:300]}  ...")

    # Step 4: JSON import + arch filter
    print("\n--- Step 4: JSON Import & Filter ---")
    imported = GroupedConvRegistry.from_json(json_str)
    print(f"  Imported: {len(imported)} kernels")
    filtered = imported.filter_by_arch(arch)
    print(f"  After arch filter ({arch}): {len(filtered)} kernels")
    fwd_only = imported.filter_by_variant("forward")
    print(f"  Forward only: {len(fwd_only)} kernels")

    # Step 5: GPU execution with a problem
    print("\n--- Step 5: GPU Execution ---")
    runner = GpuGroupedConvRunner()
    if not runner.is_available():
        print("  GPU library not available")
        return 1

    print(f"  Compiled kernels: {runner.lib.kernel_names()}")

    prob = GroupedConvProblem(
        N=1, C=128, K=128, Hi=16, Wi=16, Y=3, X=3,
        stride_h=1, stride_w=1, pad_h=1, pad_w=1,
        direction="forward",
    )
    inp = np.random.uniform(-0.3, 0.3, prob.input_shape()).astype(np.float16)
    wei = np.random.uniform(-0.3, 0.3, prob.weight_shape()).astype(np.float16)

    res = runner.run(inp, wei, prob)
    if res.success:
        print(f"  Time:   {res.time_ms:.4f} ms")
        print(f"  TFLOPS: {res.tflops:.2f}")
        print(f"  Output: {res.output.shape}, nonzero={np.count_nonzero(res.output)}/{res.output.size}")
    else:
        print(f"  GPU failed: {res.error}")

    runner.cleanup()

    # Summary
    print("\n" + "=" * 70)
    print(f"  Registries:  throughput={len(tp_reg)}, latency={len(lat_reg)}")
    print(f"  Combined:    {len(combined)} kernels")
    print(f"  JSON:        round-trip OK ({len(imported)} imported)")
    gpu_ok = res.success if runner.is_available() else False
    print(f"  GPU:         {'OK' if gpu_ok else 'FAIL'}")
    print(f"  Status:      {'PASS' if gpu_ok else 'FAIL'}")
    print("=" * 70)
    return 0 if gpu_ok else 1


if __name__ == "__main__":
    sys.exit(main())
